# The battle of neighbourhoods - Report

## Introduction: the business problem

The client is a big hotel chain who wants to estabilish itself in the city of Munich. The current subdivision of Munich in neighbourhood, while historically grounded, fails to mirror the internal subdivision of the city. The client is interested in getting a picture of the predominant activities in the different sectors of the city, so as to understand where to open which kind of hotel: a family resort would be ideally placed in an area with parks rather than an industrial area, a more hostel-kind of accomodation close to cafes and pubs, and so on.

It is a typical clustering problem, where we are both interested in geographical proximity of points - neighbourhood should be connected - and some kind of "cultural" proximity, i.e. similar kind of activities. This report summarizes the full notebook, to which we refer for details.

## Data

We review the situation as-is, i.e. the historical neighbourhood/postal codes subdivision:

In [2]:
pc_mun = pd.read_csv("postal_codes_munich.csv")
pc_mun.head()

zipcode                 Neighbourhood  latitude  longitude
0    80331                Altstadt-Lehel   48.1345    11.5710
1    80333                Altstadt-Lehel   48.1452    11.5668
2    80333                   Maxvorstadt   48.1452    11.5668
3    80335                Altstadt-Lehel   48.1427    11.5552
4    80335  Ludwigsvorstadt-Isarvorstadt   48.1427    11.5552

It contains 25 different neighbourhoods and 74 different postal codes:

In [393]:
print("Unique neighbourhoods: " + str(len(pc_mun["Neighbourhood"].unique())))
print("Unique postal codes: " + str(len(pc_mun["zipcode"].unique())))

Unique neighbourhoods: 25
Unique postal codes: 74


The tragic fact here is that postal codes aren't a refinement of the neighbourhoods, as can already be seen from the first rows of the dataset - 80333 corresponds to both Altstadt-Lehel and Maxvorstadt, covering two areas which we would expect to see in different clusters.

So we build a grid over Munich and then cluster the points of the grid:

In [394]:
latitudes = np.linspace(start = 48.09, stop = 48.20, num = 20)
longitudes = np.linspace(start = 11.48, stop = 11.65, num = 20)

grid = pd.DataFrame(index = pd.MultiIndex.from_product([latitudes, longitudes], names = ["Latitude", "Longitude"])).reset_index()
grid["Neighbourhood"] = grid.index
grid.head()

Latitude  Longitude  Neighbourhood
0     48.09  11.480000              0
1     48.09  11.488947              1
2     48.09  11.497895              2
3     48.09  11.506842              3
4     48.09  11.515789              4

Neat, right? Now let's do the clustering.

## Methodology

We set up the classification algorithm and let it run, just like in the previous course assignments. We take around every element of the grid a circle of radius 75% the distance between the points, so that we cover all of the map, at the cost of counting some element twice.

We can load the activities for each grid point from Foursquare and obtain a dataframe like this:

In [287]:
munich_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0             0                  48.09                   11.48   
1             0                  48.09                   11.48   
2             0                  48.09                   11.48   
3             0                  48.09                   11.48   
4             0                  48.09                   11.48   

                            Venue  Venue Latitude  Venue Longitude  \
0                        Rossmann       48.087505        11.484407   
1                 Schweizer Platz       48.088626        11.479916   
2                            REWE       48.089149        11.480729   
3                   Ratschiller's       48.089207        11.480467   
4  Wochenmarkt am Schweizer Platz       48.089108        11.480147   

   Venue Category  
0       Drugstore  
1           Plaza  
2     Supermarket  
3          Bakery  
4  Farmers Market

In [288]:
print('There are {} uniques categories.'.format(len(munich_venues['Venue Category'].unique())))

There are 320 uniques categories.


And now for the clustering - we drop the Neighbourhood ID for the dataset, since we don't want the algorithm to use it:

In [366]:
# set number of clusters
kclusters = 28

munich_grouped_clustering = munich_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(munich_grouped_clustering)

array([12, 12, 12, 12, 12, 12, 12, 10, 10,  7,  7,  7], dtype=int32)

Now we have the clusters and we can merge everything back in the original dataframe. We add columns with the most common venues per grid point, which will help us analyse the results later:

In [367]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = munich_grouped['Neighborhood']

for ind in np.arange(munich_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(munich_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

munich_merged = grid

munich_merged = munich_merged.merge(neighborhoods_venues_sorted.set_index('Neighborhood'), left_on='Neighbourhood', right_on = "Neighborhood")

munich_merged.head() # check the last columns!

Latitude  Longitude  Neighbourhood  Cluster Labels 1st Most Common Venue  \
0  48.095789  11.488947             21              12     German Restaurant   
1  48.095789  11.497895             22              12           Supermarket   
2  48.095789  11.506842             23              12  Gym / Fitness Center   
3  48.095789  11.515789             24              12                 Hotel   
4  48.095789  11.524737             25              12                Bakery   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0            Playground      Italian Restaurant                Castle   
1      Asian Restaurant           Bowling Alley                Bakery   
2                 Hotel  Furniture / Home Store             Pet Store   
3                  Bank                    Café           Supermarket   
4  Gym / Fitness Center        Greek Restaurant           Supermarket   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Accessories Store     Outdoor Sculpture       Organic Grocery   
1                         Spa        Massage Studio             Drugstore   
2  Modern European Restaurant           Supermarket       Organic Grocery   
3                   Pet Store             Drugstore           Pizza Place   
4            Doner Restaurant   Rental Car Location                  Café   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Optical Shop           Opera House                 Office  
1               Metro Station             Nightclub           Optical Shop  
2                   Drugstore      Asian Restaurant      Electronics Store  
3  Construction & Landscaping      Greek Restaurant         Ice Cream Shop  
4                Noodle House       Organic Grocery           Optical Shop

## Results

Here's the number of points per cluster:

In [404]:
munich_merged.groupby("Cluster Labels").agg({"Cluster Labels":"count"})

Cluster Labels
Cluster Labels                
0                           18
1                           12
2                           14
3                            3
4                           12
5                           25
6                           17
7                           22
8                           19
9                            5
10                           3
11                           1
12                          24
13                           6
14                          13
15                          20
16                          17
17                           1
18                          10
19                           2
20                          23
21                           1
22                           4
23                          16
24                           1
25                           6
26                          11
27                          16

Of course, it's so much better to visualize this on a map - take a look at the notebook for an interactive map.

## Discussion

The city center is divided in four parts, corresponding to clusters 0, 8, 20 and 6 - let's look at them one by one as an example.

### Cluster 0 - Maxvorstadt and Schwabing West

These are the point in the cluster:

In [397]:
munich_merged.loc[munich_merged['Cluster Labels'] == 0, munich_merged.columns[list(range(2, munich_merged.shape[1]))]]

Neighbourhood  Cluster Labels  1st Most Common Venue  \
153            190               0             Restaurant   
154            191               0                   Café   
170            209               0                   Café   
171            210               0                   Café   
172            211               0                   Café   
173            212               0                   Café   
188            229               0                   Café   
189            230               0                   Café   
190            231               0                    Bar   
191            232               0         Ice Cream Shop   
206            249               0                    Bar   
207            250               0     Italian Restaurant   
208            251               0     Italian Restaurant   
209            252               0                    Bar   
225            270               0  Vietnamese Restaurant   
226            271               0  Vietnamese Restaurant   
227            272               0                    Bar   
244            291               0                  Plaza   

    2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
153                 Plaza                       Café     German Restaurant   
154                 Plaza                   Boutique     French Restaurant   
170      Asian Restaurant  Middle Eastern Restaurant               Theater   
171        History Museum                 Art Museum    Italian Restaurant   
172    Italian Restaurant             Ice Cream Shop        Breakfast Spot   
173    Italian Restaurant             Ice Cream Shop             Surf Spot   
188      Asian Restaurant                 Steakhouse                   Bar   
189                   Bar                     Bakery    Italian Restaurant   
190                  Café         Italian Restaurant        Ice Cream Shop   
191             Irish Pub                Beer Garden                  Café   
206           Supermarket                       Café     Trattoria/Osteria   
207                  Café      Vietnamese Restaurant                 Plaza   
208                  Café             Ice Cream Shop        Breakfast Spot   
209                  Café          Trattoria/Osteria    Italian Restaurant   
225   Japanese Restaurant                        Bar     Indian Restaurant   
226           Pizza Place                      Plaza           Supermarket   
227    Italian Restaurant            Organic Grocery                Bakery   
244                  Café                   Bus Stop                Bakery   

        5th Most Common Venue        6th Most Common Venue  \
153                    Church                 Tram Station   
154              Cocktail Bar               Clothing Store   
170            History Museum                   Restaurant   
171       Japanese Restaurant                        Plaza   
172                       Bar                 Cocktail Bar   
173                     River  Eastern European Restaurant   
188                    Bakery            German Restaurant   
189  Mediterranean Restaurant        Vietnamese Restaurant   
190          Asian Restaurant            German Restaurant   
191                Restaurant                 Optical Shop   
206                      Pool          Bavarian Restaurant   
207          Greek Restaurant                          Bar   
208     Vietnamese Restaurant                  Pizza Place   
209              Burger Joint            German Restaurant   
225               Supermarket           Mexican Restaurant   
226           Thai Restaurant            Indian Restaurant   
227                      Café                 Burger Joint   
244                   Taverna                    Drugstore   

    7th Most Common Venue     8th Most Common Venue 9th Most Common Venue  \
153      Botanical Garden                 Nightclub         Shopping Mall   
154                   Bar    

This is where we'd build the youth hostel - look at the amount of cafes and bars in the top three. Maxvorstadt and Schwabing are indeed known to be hip neighbourhoods.

### Cluster 8 - Altstadt-Lehel and Au-Haidhausen

These are the points in the cluster:

In [398]:
munich_merged.loc[munich_merged['Cluster Labels'] == 8, munich_merged.columns[list(range(2, munich_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  \
101            132               8                 Plaza   
102            133               8    Italian Restaurant   
103            134               8                 Hotel   
104            135               8              Bus Stop   
118            151               8                  Café   
119            152               8     Indian Restaurant   
120            153               8    Italian Restaurant   
121            154               8     German Restaurant   
122            155               8    Italian Restaurant   
136            171               8                  Café   
137            172               8     Indian Restaurant   
138            173               8                  Café   
139            174               8     Indian Restaurant   
140            175               8             Drugstore   
155            192               8    Italian Restaurant   
156            193               8     German Restaurant   
157            194               8    Italian Restaurant   
174            213               8    Italian Restaurant   
175            214               8                 Plaza   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
101                 Hotel           Burger Joint                Bakery   
102                 Plaza      French Restaurant            Restaurant   
103  Gym / Fitness Center           Climbing Gym                   Pub   
104                   Pub         Discount Store        Shipping Store   
118   Bavarian Restaurant            Coffee Shop          Cocktail Bar   
119        Ice Cream Shop            Pizza Place          Concert Hall   
120                  Café                  Plaza                Bakery   
121                 Hotel     Italian Restaurant                  Café   
122                 Hotel  Portuguese Restaurant     Indian Restaurant   
136   Bavarian Restaurant            Coffee Shop                 Hotel   
137                 Beach     Italian Restaurant           Supermarket   
138                 Plaza      French Restaurant           Beer Garden   
139    Italian Restaurant            Supermarket          Burger Joint   
140  Gym / Fitness Center            Supermarket      Greek Restaurant   
155   Japanese Restaurant      German Restaurant                 Hotel   
156    Italian Restaurant             Art Museum            Restaurant   
157            Restaurant           Tram Station                 Plaza   
174     German Restaurant                  River           Supermarket   
175                Bakery     Italian Restaurant     German Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
101           Beer Garden           Supermarket      Greek Restaurant   
102                  Café           Supermarket    Turkish Restaurant   
103             Nightclub             Beach Bar              Beer Bar   
104             Beach Bar          Liquor Store             Nightclub   
118           Pizza Place             Bookstore     German Restaurant   
119        Science Museum           Supermarket                 Hotel   
120     German Restaurant     Indian Restaurant                   Bar   
121                 Plaza     Indian Restaurant    Spanish Restaurant   
122          Home Service           Pizza Place      Doner Restaurant   
136                 Plaza             Bookstore           Pizza Place   
137                Bakery           Beer Garden           Snack Place   
138        Ice Cream Shop    Italian Restaurant     German Restaurant   
139  Gym / Fitness Center    Mexican Restaurant      Greek Restaurant   
140      Sushi Restaurant                   Bar          Gourmet Shop   
155                   Bar          Cocktail Bar                 Plaza   
156     Indian Restaurant          Tram Station   Monument / Landmark   
157                  Café     Indian Restaurant                  Pool   
174      

This is a more residential area - we often see drugstores and supermarkets in the first positions. It's still central, but looking at the amount of restaurants vs bars we can tell that the target is different. Hotels are also more common.

### Cluster 20 - Ludwigsvorstadt-Isarvorstadt and Sendling

These are the points in the cluster:

In [399]:
munich_merged.loc[munich_merged['Cluster Labels'] == 20, munich_merged.columns[list(range(2, munich_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  \
43              68              20                 Plaza   
44              69              20                  Park   
61              88              20                 Plaza   
62              89              20    Athletics & Sports   
63              90              20               Taverna   
78             107              20     German Restaurant   
79             108              20    Italian Restaurant   
80             109              20                   Bar   
81             110              20     German Restaurant   
82             111              20    Italian Restaurant   
94             125              20    Basketball Stadium   
95             126              20                  Park   
96             127              20    Seafood Restaurant   
97             128              20    Italian Restaurant   
98             129              20    Italian Restaurant   
99             130              20                   Bar   
100            131              20    Italian Restaurant   
113            146              20    Italian Restaurant   
114            147              20                  Café   
115            148              20    Italian Restaurant   
116            149              20    Italian Restaurant   
117            150              20                  Café   
135            170              20                  Café   

     2nd Most Common Venue  3rd Most Common Venue       4th Most Common Venue  \
43      Turkish Restaurant           Soccer Field  Construction & Landscaping   
44                   Beach              Gastropub                        Café   
61      Italian Restaurant      German Restaurant          Turkish Restaurant   
62              Restaurant                   Park                       Trail   
63                   Plaza              Drugstore                    Bus Line   
78        Doner Restaurant                   Bank                   Gastropub   
79             Supermarket                 Market           Food & Drink Shop   
80                    Café     Italian Restaurant                 Supermarket   
81               Drugstore           Soccer Field                       Plaza   
82                     Bar                  Plaza                 Pizza Place   
94                  Tunnel           Tennis Court                      Lounge   
95              Playground     Italian Restaurant                        Café   
96                    Café  Vietnamese Restaurant           Indian Restaurant   
97                    Café     Seafood Restaurant               Grocery Store   
98                    Café                    Bar                      Bakery   
99                    Café     Italian Restaurant            Asian Restaurant   
100                   Café                    Pub                    Bus Stop   
113                   Café            Music Venue                    Bus Stop   
114      Indian Restaurant       Asian Restaurant         Bavarian Restaurant   
115           Burger Joint                   Café               Grocery Store   
116  Vietnamese Restaurant                   Café           German Restaurant   
117            Coffee Shop                    Bar          Italian Restaurant   
135     Italian Restaurant                  Plaza           German Restaurant   

     5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
43                   Hotel                 Bakery           Beer Garden   
44               Rest Area     Seafood Restaurant           Beer Garden   
61             Gas Station                 Bakery                  Park   
62             Gas Station           Soccer Field                 Beach   
63                     Spa               Bus Stop                  Café   
78                Bus Stop  Vietnamese Restaurant                  Café   
79               Gastropub          Grocery Store    Falafel Restaurant   
80      Turkish Res

Yet another cafe area - it probably got separated from Cluster 0 because the algorithm tends to prefer circular neighbourhoods.

### Cluster 6 - Schwantalerhöhe and Neuhausen

In [409]:
munich_merged.loc[munich_merged['Cluster Labels'] == 18, munich_merged.columns[list(range(2, munich_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  \
192            233              18              Bus Stop   
193            234              18                  Bank   
210            253              18                   Bar   
211            254              18          Bathing Area   
212            255              18          Bathing Area   
228            273              18                 Hotel   
229            274              18    Photography Studio   
231            276              18              Bus Stop   
247            294              18               Stadium   
248            295              18         Indie Theater   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
192   Monument / Landmark            Hotel Pool   Bavarian Restaurant   
193                Bakery           Supermarket                  Park   
210                Tunnel               Dog Run           Comedy Club   
211   Bavarian Restaurant          Tennis Court                  Café   
212     Trattoria/Osteria              Bus Stop     Accessories Store   
228     German Restaurant                   Bar     Afghan Restaurant   
229          Bathing Area   Bavarian Restaurant                  Park   
231      Asian Restaurant           Gas Station                Bakery   
247       Bed & Breakfast                 Trail           Beer Garden   
248                  Lake                  Park          Bathing Area   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
192            Restaurant           Beer Garden           Snack Place   
193          Gourmet Shop    Athletics & Sports                Hostel   
210     Trattoria/Osteria           Boat Rental     German Restaurant   
211     Accessories Store          Noodle House       Organic Grocery   
212             Newsstand       Organic Grocery          Optical Shop   
228              Bus Stop     Trattoria/Osteria             Nightclub   
229          Tennis Court      Volleyball Court           Beer Garden   
231               Theater                   Bar    Italian Restaurant   
247         Moving Target             Newsstand          Optical Shop   
248               Dog Run            Skate Park           Music Venue   

    8th Most Common Venue    9th Most Common Venue 10th Most Common Venue  
192     Recreation Center       Athletics & Sports                  Hotel  
193     German Restaurant          Organic Grocery               Bus Stop  
210           Snack Place        Convenience Store            Beer Garden  
211          Optical Shop              Opera House                 Office  
212           Opera House                   Office           Noodle House  
228          Outlet Store        Outdoor Sculpture        Organic Grocery  
229             Newsstand              Opera House                 Office  
231           Music Store              Opera House           Outlet Store  
247              Mountain              Opera House                 Office  
248       Nature Preserve  New American Restaurant              Newsstand

This is the area around the central station - look at how many hotels!

### Other clusters

The algorithm caught some other interesting features, for example the industrial area northwest (Cluster 9), the zoo in the south (Cluster 10) and the park (Cluster 18).

## Conclusion

The algorithm provides a new rationale for the subdivision of the city, grouping and splitting old neighbourhoods into new ones.